colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py

```diff
@@ -0,0 +1,378 @@
+                            cost_table[m][i][idx] = best_leaf[1]
+                            back_ptr[m][i][idx] = (False, best_leaf[0])
+                        else:
+                            cost_table[m][i][idx] = chain_checkpoint
+                            back_ptr[m][i][idx] = (True,)
+        return cost_table, back_ptr
+
+    @staticmethod
+    def _compute_table_c(chain: Chain, mem_slots: int) -> Tuple:
+        raise NotImplementedError("C implementation not available yet")
+
+    def _backtrack(self, chain: Chain, lmin: int, lmax: int, mem_budget: int, cost_table: List[List[Dict[int, Tuple]]],
+                   back_ptr: List[List[Dict[int, int]]]) -> List[int]:
+        """Backtrack the cost table and retrieve the optimal checkpointing strategy.
+
+        Args:
+            chain (Chain): A basic linearized structure for solving the dynamic programming problem.
+            lmin (int): The left index of the interval to backtrack.
+            lmax (int): The right index of the interval to backtrack.
+            mem_budget (int): The memory budget for processing this interval.
+            cost_table (List[List[Dict[int, Tuple]]]): See _compute_table() for definitions
+            back_ptr (List[List[Dict[int, Tuple]]]): See _compute_table() for definitions
+
+        Raises:
+            ValueError: Can not process the chain.
+
+        Returns:
+            sequence (Sequence): The sequence of executing nodes with checkpoints.
+        """
+        if mem_budget <= 0:
+            raise ValueError(f"Can not process a chain with negative memory {mem_budget}")
+        elif cost_table[mem_budget][lmin][lmax] == float("inf"):
+            raise ValueError(f"Can not process this chain from index {lmin} to {lmax} with memory {mem_budget}")
+
+        sequence = Sequence(Function("Persistent", lmax - lmin, mem_budget))
+        if lmin == lmax:
+            if lmin == chain.length:
+                sequence.insert(Loss())
+            else:
+                sequence.insert(ForwardEnable(lmin))
+                sequence.insert(Backward(lmin))
+            return sequence
+
+        if back_ptr[mem_budget][lmin][lmax][0]:
+            sequence.insert(ForwardEnable(lmin))
+            sequence.insert_sequence(
+                self._backtrack(chain, lmin + 1, lmax, mem_budget - chain.xbar[lmin + 1], cost_table, back_ptr))
+            sequence.insert(Backward(lmin))
+        else:
+            j = back_ptr[mem_budget][lmin][lmax][1]
+            sequence.insert(ForwardCheck(lmin))
+            for k in range(lmin + 1, j):
+                sequence.insert(ForwardNograd(k))
+            sequence.insert_sequence(self._backtrack(chain, j, lmax, mem_budget - chain.xbar[j], cost_table, back_ptr))
+            sequence.insert_sequence(self._backtrack(chain, lmin, j - 1, mem_budget, cost_table, back_ptr))
+        return sequence
+
+    @staticmethod
+    def _annotate_from_sequence(sequence: Sequence, node_list: List[List[Node]]):
```

Comment:
is this function concise?

Reply:
This function might be more concise, I think this part could be left for me to refactor as I might need to refactor the activation codegen

Reply:
Our codegen is too heavy, we might dive deep later.